File: 39combinationSum/combinationSum.java
import java.util.ArrayList;



List<List<Integer>> allResultList = new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        if (target <= 0) {
            allResultList.add(new ArrayList<>());
        }
        //先排序，排序后可以加剪枝
        Arrays.sort(candidates);
        dfs(new ArrayList<>(), candidates, 0, target, 0);
        return allResultList;
    }


    public void dfs(List<Integer> list, int[] candidates, int sum, int target, int start) {
        for (int i = start; i < candidates.length; i++) {
            list.add(candidates[i]);
            if ((sum + candidates[i]) == target) {
                //此时tmpList 满足
                List<Integer> tmpList = new ArrayList<>(list);
                allResultList.add(tmpList);
            } if ((sum + candidates[i]) < target) {
                dfs(list, candidates, sum + candidates[i], target, i);
            } else {
                list.remove(list.size() - 1);
                //(sum+candidates[i]) > target,因为数组有序，后面一定不满足
                break;
            }
            list.remove(list.size() - 1);
        }
    }

File: 39combinationSum/combinationSum.go
package main

import (
	"fmt"
	"sort"
)

/*
  给定一个无重复元素的数组 candidates 和一个目标数 target ，找出 candidates 中所有可以使数字和为 target 的组合。
  candidates 中的数字可以无限制重复被选取。
  说明：
    所有数字（包括 target）都是正整数。
    解集不能包含重复的组合。
  示例 1:
  输入: candidates = [2,3,6,7], target = 7,
  所求解集为:
  [
  [7],
  [2,2,3]
  ]
  示例 2:
  输入: candidates = [2,3,5], target = 8,
  所求解集为:
  [
  [2,2,2,2],
  [2,3,3],
  [3,5]
  ]
*/

func main() {
	candidates := []int{2, 3, 6, 7}
	target := 7
	res := combinationSum(candidates, target)
	fmt.Println(res)
}

func combinationSum(candidates []int, target int) [][]int {
	res := make([][]int, 0)
	if len(candidates) == 0 {
		return res
	}
	path := make([]int, 0)
	sort.Ints(candidates) // 通过提交可以知道，排序可以加快速度
	combinationHelper(candidates, path, target, 0, &res)
	return res
}

func combinationHelper(candidates, path []int, target, left int, res *[][]int) {
	if target == 0 {
		fmt.Println("path is ", path)
		*res = append(*res, append([]int{}, path...))
		return
	}

	for i := left; i < len(candidates); i++ {
		if target < candidates[i] {
			return // 因为已排序的了
		}
		path = append(path, candidates[i])
		combinationHelper(candidates, path, target-candidates[i], i, res)
		path = path[:len(path)-1]
		// path也可以在下一层再加，这样就不用回漱
		// combinationHelper(candidates, append(path, candidates[i]), target-candidates[i], i, res)
	}
}

func sumSlice(res []int) int {
	s := 0
	for _, v := range res {
		s += v
	}
	return s
}
